File: sol/solution/0001-0100/0037/Solution.dart
class Solution {
  void solveSudoku(List<List<String>> board) {
    bool isValid(int r, int c, String ch) {
      for (int i = 0; i < 9; i++) {
        if (board[r][i] == ch || board[i][c] == ch) return false;
        if (board[(r ~/ 3) * 3 + (i ~/ 3)][(c ~/ 3) * 3 + (i % 3)] == ch)
          return false;
      }
      return true;
    }
    bool solve() {
      for (int i = 0; i < 9; i++) {
        for (int j = 0; j < 9; j++) {
          if (board[i][j] == ".") {
            for (int d = 1; d <= 9; d++) {
              String ch = d.toString();
              if (isValid(i, j, ch)) {
                board[i][j] = ch;
                if (solve()) return true;
                board[i][j] = ".";
              }
            }
            return false;
          }
        }
      }
      return true;
    }
    solve();
  }
}

File: sol/solution/0001-0100/0037/Solution.kt
class Solution {
    fun solveSudoku(board: Array<CharArray>): Unit {
        solve(board)
    }
    private fun solve(b: Array<CharArray>): Boolean {
        for (i in 0 until 9) {
            for (j in 0 until 9) {
                if (b[i][j] == '.') {
                    for (d in '1'..'9') {
                        if (isValid(b, i, j, d)) {
                            b[i][j] = d
                            if (solve(b)) return true
                            b[i][j] = '.'
                        }
                    }
                    return false
                }
            }
        }
        return true
    }
    private fun isValid(b: Array<CharArray>, r: Int, c: Int, ch: Char): Boolean {
        for (i in 0 until 9) {
            if (b[r][i] == ch || b[i][c] == ch) return false
            if (b[(r/3)*3 + i/3][(c/3)*3 + i%3] == ch) return false
        }
        return true
    }
}